libcontainer/cgroups/fs/devices.go

```diff
@@ -32,29 +40,63 @@ func (s *DevicesGroup) Set(path string, cgroup *configs.Cgroup) error {
 	}
 
 	devices := cgroup.Resources.Devices
+	oldAllowedDevices, err := readDevicesExceptDefault(path)
+	if err != nil {
+		return err
+	}
+
 	if len(devices) > 0 {
 		for _, dev := range devices {
 			file := "devices.deny"
 			if dev.Allow {
 				file = "devices.allow"
 			}
-			if err := fscommon.WriteFile(path, file, dev.CgroupString()); err != nil {
-				return err
+
+			// For the second time set, we don't deny all devices
+			if dev.Type == defaultDevice.Type && len(oldAllowedDevices) != 0 {
+				file = ""
+			}
```

Comment:
I _really_ don't like this, and after looking at it for a bit I think I'm convinced it isn't right. This code will ignore any update that contains `a` -- but there are valid use-cases for `a` other than "deny everything" (you can allow both character and block devices for a particular major number, for instance). But even if this check was `dev == defaultDevice` it'd still be wrong since you can also allow all devices (though you shouldn't) -- and that would be ignored.

I think the issue is that the caller shouldn't be specifying a default deny rule in the list which we then need to filter out -- that should be done as part of this module. So we figure out what rules we need to specify.